packages/flutter/lib/src/widgets/spell_check.dart

```diff
@@ -0,0 +1,322 @@
+  /// specified overrides.
+  SpellCheckConfiguration copyWith({
+    SpellCheckService? spellCheckService,
+    SpellCheckSuggestionsHandler? spellCheckSuggestionsHandler,
+    TextStyle? misspelledTextStyle}) {
+    return SpellCheckConfiguration(
+      spellCheckService: spellCheckService ?? this.spellCheckService,
+      spellCheckSuggestionsHandler:
+        spellCheckSuggestionsHandler ?? this.spellCheckSuggestionsHandler,
+      misspelledTextStyle: misspelledTextStyle ?? this.misspelledTextStyle,
+    );
+  }
+}
+
+/// Determines how misspelled words are indicated in text input and how
+/// replacement suggestions for misspelled words are displayed via menu.
+mixin SpellCheckSuggestionsHandler {
+  /// Builds the [TextSpan] tree given the current state of the text input and
+  /// spell check results.
+  TextSpan buildTextSpanWithSpellCheckSuggestions(
+    TextEditingValue value,
+    bool composingWithinCurrentTextRange,
+    TextStyle? style,
+    TextStyle misspelledTextStyle,
+    SpellCheckResults spellCheckResults
+  );
+}
+
+/// The handler used by default for spell checking text input.
+///
+/// Any widget may use this handler to build a [TextSpan] tree with
+/// [SpellCheckResults] indicated in a style based on the platform by calling
+/// `buildTextSpanWithSpellCheckSuggestions(...)` with an instance of this
+/// class.
+///
+/// See also:
+///
+///  * [SpellCheckSuggestionsHandler], the handler that this implements and
+///    may be overriden for use by [EditableText].
+///  * [EditableText], which uses this handler to display spell check results
+///     by default if spell check is enabled.
+class DefaultSpellCheckSuggestionsHandler with SpellCheckSuggestionsHandler {
+  /// Creates a handler to use for spell checking text input based on the
+  /// provided platform.
+  DefaultSpellCheckSuggestionsHandler() {}
+
+  /// Adjusts spell check results to correspond to [newText] if the only results
+  /// that the handler has access to are the [results] corresponding to
+  /// [resultsText].
+  ///
+  /// Used in the case where the request for the spell check results of the
+  /// [newText] is lagging in order to avoid display of incorrect results.
+  static List<SuggestionSpan> correctSpellCheckResults(
+      String newText, String resultsText, List<SuggestionSpan> results) {
+    final List<SuggestionSpan> correctedSpellCheckResults = <SuggestionSpan>[];
+
+    int spanPointer = 0;
+    int offset = 0;
+    int foundIndex;
+    int spanLength;
+    SuggestionSpan currentSpan;
+    SuggestionSpan adjustedSpan;
+    String currentSpanText;
+    String newSpanText = '';
+    bool currentSpanValid = false;
+    bool foundCurrentSpan = false;
+    RegExp regex;
+
+    // Assumes that the order of spans has not been jumbled for optimization
+    // purposes, and will only search since the previously found span.
+    int searchStart = 0;
+
+    while (spanPointer < results.length) {
```

Comment:
I'm not entirely following what this loop is trying to do. Maybe we can sync up offline to talk about it? @camsim99